### PR TITLE
Move Device and DeviceType to c10

### PR DESCRIPTION
**Pull request by ezyang**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#12995 Move Device and DeviceType to c10**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10513246/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13019 Replace direct use of int32_t with an alias DeviceIndex&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10520295/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #13021 Delete default constructor from CUDAStream.&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D10520421/)

Differential Revision: [D10513246](https://our.internmc.facebook.com/intern/diff/D10513246/)

### CODE REVIEW

**Review by li-roy — APPROVED** [review]
looks good, please fix builds :)

File: aten/src/ATen/Device.h [hunk 1]
```diff
@@ -1,2 +1,2 @@
 #pragma once
-#include <ATen/core/Device.h>
+#include <c10/Device.h>
```
Comment (Yangqing) [inline aten/src/ATen/Device.h]: Quick question: should we put it to c10/core?

File: c10/DeviceType.cpp [hunk 2]
```diff
@@ -1,9 +1,9 @@
-#include <ATen/core/DeviceType.h>
+#include <c10/DeviceType.h>
 #include <c10/util/Exception.h>
 
-namespace at {
+namespace c10 {
 
-std::string DeviceTypeName(at::DeviceType d, bool lower_case) {
+std::string DeviceTypeName(DeviceType d, bool lower_case) {
```
Comment (Yangqing) [inline c10/DeviceType.cpp]: nit: in line 27-30, we still mention caffe2.proto which we might want to just remove? :)

File: c10/DeviceType.h [hunk 3]
```diff
@@ -26,18 +26,24 @@ enum class DeviceType : int32_t {
...
-template <> struct hash<at::DeviceType> {
-  std::size_t operator()(const at::DeviceType &k) const {
+template <> struct hash<c10::DeviceType> {
+  std::size_t operator()(c10::DeviceType k) const {
     return std::hash<int>()(static_cast<int>(k));
   }
 };
 } // namespace std
+
+// TODO: Remove me when we get a global c10 namespace using in at
```
Comment (Yangqing) [inline c10/DeviceType.h]: thumbs up

**Review by Yangqing — APPROVED** [review]
Thanks so much! Approved with minor comments.

File: c10/Device.h [hunk 4]
```diff
@@ -110,3 +110,8 @@ struct hash<at::Device> {
   }
 };
 } // namespace std
+
+// TODO: Remove when we add global namespace include
+namespace at {
+using c10::Device;
```
Comment (Yangqing) [inline c10/Device.h]: I don't quite get it why, but windows build failed with this:

19:01:37 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\c10/Device.h(116): error C2874: using-declaration causes a multiple declaration of 'c10::Device'
19:01:37 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\c10/Device.h(24): note: see declaration of 'c10::Device'

File: aten/src/ATen/Device.h (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/Device.h]: I am on the record for wanting user-visible classes to live at the top-level!

Comment (Yangqing) [reply aten/src/ATen/Device.h]: Sounds good to me ;)

**Pull request closed without merging**